drivers/serial/uart_hostlink.c

```diff
@@ -0,0 +1,410 @@
+
+	compiler_barrier();
+}
+
+/*
+ * Wait for host response and return pointer to hostlink payload.
+ * Symbol hl_blockedPeek() is used by the simulator as message signal.
+ */
+static void __noinline _hl_blockedPeek(void)
+{
+	while (hl_read32(&__HOSTLINK__.hdr.host2target_addr) == HL_NOADDRESS) {
+		/* TODO: Timeout.  */
+	}
+}
+
+static void hl_static_recv(void)
+{
+	compiler_barrier();
+	_hl_blockedPeek();
+	compiler_barrier();
+}
+
+/* Mark hostlink buffer as "No message here".  */
+static void hl_delete(void)
+{
+	hl_write32(&__HOSTLINK__.hdr.target2host_addr, HL_NOADDRESS);
+}
+
+/* Parameter types.  */
+#define PAT_CHAR	1
+#define PAT_SHORT	2
+#define PAT_INT		3
+#define PAT_STRING	4
+/* For future use.  */
+#define PAT_INT64	5
+
+static void hl_static_pack_int(volatile struct hl_packed_int *pack, int32_t value)
+{
+	hl_write16(&pack->type, PAT_INT);
+	hl_write16(&pack->size, 4);
+	hl_write32(&pack->value, value);
+}
+
+static void hl_static_pack_char(volatile struct hl_packed_short_buff *pack, unsigned char c)
+{
+	hl_write16(&pack->type, PAT_STRING);
+	hl_write16(&pack->size, 1);
+	hl_write8(&pack->payload_short, c);
+}
+
+static int hl_static_unpack_int(volatile struct hl_packed_int *pack, int32_t *value)
+{
+	uint16_t type = hl_read16(&pack->type);
+	uint16_t size = hl_read16(&pack->size);
+
+	if (type != PAT_INT) {
+		return -1;
+	}
+
+	if (size != 4)  {
+		return -1;
+	}
+
+	*value = hl_read32(&pack->value);
+
+	return 0;
+}
+
+static inline int32_t hl_write_char(int fd, const char c)
+{
+	/*
+	 * Format:
+	 * in, int -> syscall (HL_SYSCALL_WRITE)
+	 * in, int -> file descriptor
+	 * in, ptr -> buffer
+	 * in, int -> bytes number
+	 * out, int -> bytes written
+	 * out, int, host errno
+	 */
+
+	hl_static_pack_int(&__HOSTLINK__.payload.pkt_write_char_put.syscall_nr, HL_SYSCALL_WRITE);
+
+	hl_static_pack_int(&__HOSTLINK__.payload.pkt_write_char_put.fd, fd);
+
+	hl_static_pack_char(&__HOSTLINK__.payload.pkt_write_char_put.buff, c);
+
+	hl_static_pack_int(&__HOSTLINK__.payload.pkt_write_char_put.nbyte, 1);
+
+	hl_static_send(sizeof(struct hl_pkt_write_char_put));
+	hl_static_recv();
+
+	int32_t bwr = 0;
+	int ret = hl_static_unpack_int(&__HOSTLINK__.payload.pkt_write_char_get.byte_written, &bwr);
+
+	/* we can get host errno here with:
+	 * hl_static_unpack_int(&__HOSTLINK__.pkt_write_char_get.host_errno, &host_errno);
+	 * but we don't need it for UART emulation.
+	 */
+
+	if (bwr <= 0) {
+		ret = -1;
+	}
+
+	hl_delete();
+
+	return ret;
+}
+
+static int uart_hostlink_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+/**
+ * @brief Poll the device for input.
+ *
+ * @param dev UART device struct
+ * @param c Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+static int uart_hostlink_poll_in(const struct device *dev, unsigned char *c)
+{
+	ARG_UNUSED(dev);
+
+	/* We plan to use hostlink for logging, so no much sense in poll_in implementation */
```

Comment:
If it is not too complex, why not add input? That enables shell functionality, and other simple interactivity if needed.

Reply:
As I remember shell require interrupt functionality from UART, so implementing of polling is not enough. But I'll check if it is so.

Reply:
Agreed - hostlink stdin may be blocking instead of polling, let's continue with this first.